common/errors.go

```diff
@@ -24,119 +62,169 @@ func NewError(format string, a ...interface{}) Error {
 
 }
 
-// New Error with cause where the type is the cause, with message..
-func NewErrorWithCause(cause error, format string, a ...interface{}) Error {
-	msg := Fmt(format, a...)
-	return newError(msg, cause, cause)
-}
-
 // New Error with specified type and message.
 func NewErrorWithType(type_ interface{}, format string, a ...interface{}) Error {
 	msg := Fmt(format, a...)
 	return newError(msg, nil, type_)
 }
 
-type traceItem struct {
-	msg      string
-	filename string
-	lineno   int
-}
+// NOTE: The name of a function "NewErrorWithCause()" implies that you are
+// creating a new Error, yet, if the cause is an Error, creating a new Error to
+// hold a ref to the old Error is probably *not* what you want to do.
+// So, use WrapCause(cause, format, a...) instead, which returns the same error
+// if cause is an Error.
+// IF you must set an Error as the cause of an Error,
+// then you can use the WithCauser interface to do so manually.
+// e.g. (error).(tmlibs.WithCauser).WithCause(causeError)
 
-func (ti traceItem) String() string {
-	return fmt.Sprintf("%v:%v %v", ti.filename, ti.lineno, ti.msg)
+type WithCauser interface {
+	WithCause(cause interface{}) Error
 }
 
 type cmnError struct {
-	msg    string
-	cause  error
-	type_  interface{}
-	traces []traceItem
+	msg        string         // first msg which also appears in msg
+	cause      interface{}    // underlying cause (or panic object)
+	type_      interface{}    // for switching on error
+	msgtraces  []msgtraceItem // all messages traced
+	stacktrace []uintptr      // first stack trace
 }
 
-// NOTE: Do not expose, it's not very friendly.
-func newError(msg string, cause error, type_ interface{}) *cmnError {
+var _ WithCauser = &cmnError{}
+var _ Error = &cmnError{}
+
+// NOTE: do not expose.
+func newError(msg string, cause interface{}, type_ interface{}) *cmnError {
 	return &cmnError{
-		msg:    msg,
-		cause:  cause,
-		type_:  type_,
-		traces: nil,
+		msg:        msg,
+		cause:      cause,
+		type_:      type_,
+		msgtraces:  nil,
+		stacktrace: nil,
 	}
 }
 
+func (err *cmnError) Message() string {
+	return err.msg
+}
+
 func (err *cmnError) Error() string {
-	return fmt.Sprintf("Error{%v:%s,%v,%v}", err.type_, err.msg, err.cause, len(err.traces))
+	return fmt.Sprintf("%v", err)
+}
+
+// Captures a stacktrace if one was not already captured.
+func (err *cmnError) Stacktrace() Error {
+	if err.stacktrace == nil {
+		err.stacktrace = captureStacktrace(3, 32)
```

Comment:
magic numbers

Reply:
the 3 I think arguably should be magic, though it deserves comments.  I'll move the 32 out.

Reply:
you don't have to move them out, you can do

```
var offset = 3
const defaultDepth = 32
```

inside the function